features/dd-sdk-android-trace/src/test/kotlin/com/datadog/android/trace/OtelTracerBuilderProviderTest.kt

```diff
@@ -446,4 +421,131 @@ internal class OtelTracerBuilderProviderTest {
+
+    @Test
+    fun `M use the sample rate W setSampleRate`(@DoubleForgery(min = 0.0, max = 100.0) sampleRate: Double) {
+        // Given
+        val expectedNormalizedSampleRate = sampleRate / 100.0
+        val tracer = testedOtelTracerProviderBuilder.setSampleRate(sampleRate).build()
+            .tracerBuilder(fakeInstrumentationName).build()
+
+        // When
+        val coreTracer: CoreTracer = tracer.getFieldValue("tracer")
+
+        // Then
+        val config: Config = coreTracer.getFieldValue("initialConfig")
+        assertThat(config.traceSampleRate).isCloseTo(expectedNormalizedSampleRate, offset(0.005))
+    }
+
+    @Test
+    fun `M use user-keep priority W buildSpan { provided keep sample rate }`() {
+        // Given
+        val tracer = testedOtelTracerProviderBuilder
+            .setPartialFlushThreshold(1)
+            .setSampleRate(100.0)
+            .build()
+            .tracerBuilder(fakeInstrumentationName)
+            .build()
+
+        // When
+        val span = tracer
+            .spanBuilder(fakeOperationName)
+            .startSpan()
+        val delegateSpan: DDSpan = span.getFieldValue("delegate")
+        delegateSpan.forceSamplingDecision()
+        span.end()
+
+        // Then
+        val priority = delegateSpan.samplingPriority
+        assertThat(priority).isEqualTo(PrioritySampling.USER_KEEP.toInt())
+    }
+
+    @Test
+    fun `M use user-drop priority W buildSpan { provide not keep sample rate }`() {
+        // Given
+        val tracer = testedOtelTracerProviderBuilder
+            .setPartialFlushThreshold(1)
+            .setSampleRate(0.0)
+            .build()
+            .tracerBuilder(fakeInstrumentationName)
+            .build()
+
+        // When
+        val span = tracer
+            .spanBuilder(fakeOperationName)
+            .startSpan()
+        val delegateSpan: DDSpan = span.getFieldValue("delegate")
+        delegateSpan.forceSamplingDecision()
+        span.end()
+
+        // Then
+        val priority = delegateSpan.samplingPriority
+        assertThat(priority).isEqualTo(PrioritySampling.USER_DROP.toInt())
+    }
+
+    @Test
+    fun `M use user-keep or user-not-keep priority W buildSpan { provided random sample rate }`(
+        @DoubleForgery(min = 0.0, max = 100.0) sampleRate: Double
+    ) {
+        // Given
+        val tracer = testedOtelTracerProviderBuilder
+            .setPartialFlushThreshold(1)
+            .setSampleRate(sampleRate)
+            .build()
+            .tracerBuilder(fakeInstrumentationName)
+            .build()
+
+        // When
+        val span = tracer
+            .spanBuilder(fakeOperationName)
+            .startSpan()
+        val delegateSpan: DDSpan = span.getFieldValue("delegate")
+        delegateSpan.forceSamplingDecision()
+        span.end()
+
+        // Then
+        val priority = delegateSpan.samplingPriority
+        assertThat(priority).isIn(PrioritySampling.USER_DROP.toInt(), PrioritySampling.USER_KEEP.toInt())
```

Comment:
> [!CAUTION]
> This part still doesn't test that the appropriate sample rate is being used… the fact that a specific field is set on the core tracer (as tested line 445) doesn't mean it's the one that will be applied. Since we're testing the actual priority set on the spans, we should really try creating 100 spans and measure the effective sampling rate and compare that to the expected one (with some margin of error)